tests/test_permissions.py

```diff
@@ -0,0 +1,451 @@
+                ActionUsers.allow
+                if allow_deny == "allow"
+                else ActionUsers.deny
+            )
+        elif isinstance(user_role, Need) and user_role.method == "system_role":
+            # SystemRoleNeed
+            kwargs = dict(role=user_role)
+            method = (
+                ActionSystemRoles.allow
+                if allow_deny == "allow"
+                else ActionSystemRoles.deny
+            )
+        else:
+            # Role
+            kwargs = dict(role=user_role)
+            method = (
+                ActionRoles.allow
+                if allow_deny == "allow"
+                else ActionRoles.deny
+            )
+
+        obj = method(action=action_need, **kwargs, argument=argument)
+        db.session.add(obj)
+    db.session.commit()
+    return action_need
+
+
+def test_allow_by_user_id(access_app):
+    """Ensure allow permission by user id."""
+    admin, reader = create_users("admin", "reader")
+
+    permission, = create_permissions({"needs": {UserNeed(admin.id)}})
+
+    assert permission.allows(admin)
+    assert permission.allows(get_superuser())
+    assert not permission.allows(reader)
+
+
+def test_deny_by_user_id(access_app):
+    """Ensure deny permission when user is denied."""
+    admin, reader = create_users("admin", "reader")
+
+    permission, = create_permissions(
+        {"excludes": [UserNeed(admin.id), UserNeed(reader.id)]}
+    )
+
+    assert not permission.allows(admin)
+    assert not permission.allows(reader)
+    assert permission.allows(get_superuser())
+
+
+def test_allow_deny_by_user_id(access_app):
+    """Ensure allow/deny permission when same user is allowed and denied."""
+    admin, reader = create_users("admin", "reader")
+
+    permission, = create_permissions(
+        {
+            "needs": [UserNeed(admin.id)],
+            "excludes": [UserNeed(admin.id), UserNeed(reader.id)],
+        }
+    )
+
+    assert not permission.allows(admin)
+    assert not permission.allows(reader)
+    assert permission.allows(get_superuser())
+
+
+def test_allow_by_role(access_app):
+    """Ensure allow permission by role."""
+    admin, reader = create_users("admin", "reader")
+    administrators, readers = create_roles("administrators", "readers")
+    assign_roles({admin: [administrators], reader: [readers]})
+
+    permission, = create_permissions(
+        {"needs": [RoleNeed(administrators.name)]}
+    )
+
+    assert permission.allows(admin)
+    assert not permission.allows(reader)
+    assert permission.allows(get_superuser())
+
+
+def test_deny_by_role(access_app):
+    """Ensure deny permission by role."""
+    admin, reader = create_users("admin", "reader")
+    administrators, readers = create_roles("administrators", "readers")
+    assign_roles({admin: [administrators], reader: [readers]})
+
+    permission, = create_permissions(
+        {
+            "needs": [RoleNeed(administrators.name)],
+            "excludes": [
```

Comment:
Minor: Nice! You can add a comment explaining that `excludes` prevail over `needs`. It might be confusing otherwise.